/*
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        https://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

syntax = "proto3";
import "google/protobuf/any.proto";
import "SerializableDataType.proto";
package NES;


message SchemaIdentifier{
  string name = 1;
  optional string table = 2;
}
/*
The serializable wrapper definition for fields
 */
message SerializableField{
  SchemaIdentifier identifier = 1;
  SerializableDataType type = 2;
}

/**
The serializable wrapper definition for function nodes
 */
message SerializableFunction {
  message ConstantExpression{
    string value = 1;
  }
  message FunctionExpression{
    string functionName = 1;
  }
  message FunctionFieldAccess{
    SchemaIdentifier field = 1;
  }

  optional SerializableDataType stamp = 1;
  google.protobuf.Any details = 2;
  repeated SerializableFunction children = 3;

}

